### ISSUE TITLE
The ObjectStateMatcher should reveal the value before comparing it

### ISSUE THREAD

**Issue by stof** [phpspec/prophecy#69]
The following code will not produce the expected behavior:

```php
$double = $prophet->prophesize('...');

$double->someAction(Argument::which('getSubject', $double))->shouldBeCalled();
```

the object state matcher will not match here, because it will compare the return value of ``getSubject()`` with ``$double``, which is the ObjectProphecy instance.
To make it work, we need to reveal the double explicitly before building the matcher:

```php
$double = $prophet->prophesize('...');

$double->someAction(Argument::which('getSubject', $double->reveal()))->shouldBeCalled();
```

this is inconsistent with the implicit revealing done for calls like this:

```php
$double = $prophet->prophesize('...');

$double->someAction($double)->shouldBeCalled();
```

**Comment by mbrodala** [comment]
I think the latter case should not implicitly reveal the `$double` either since `someAction` might interact with it and thus unexpectedly add more predictions.

**Comment by stof** [comment]
@mbrodala I don't understand what you mean. ``$double->someAction($double)->shouldBeCalled()`` is not calling ``someAction``. It is defining a prediction for it.

**Comment by mbrodala** [comment]
You're right, got confused myself. But I think it should be `$double->someAction($double->reveal())->shouldBeCalled();` instead, since it is very unlikely that the prophecy will be passed to `someAction` instead of the double.

**Comment by stof** [comment]
@mbrodala being able to use the prophecy object when defining predictions will not be removed. It would be a BC break (and so cannot be done), and would force making the code less readable. note that passing the revealed double is already possible

**Comment by mbrodala** [comment]
Sure thing, just staging my POV. :-)

**Comment by jon-acker** [comment]
@stof I had a chance to look at the code, and it seems to me that in ObjectStateToken::scoreArgument, you'd need something like
```php
$value = ($this->value instanceof ProphecyInterface) ? $this->value->reveal() : $this->value;

try {
    $comparator->assertEquals($actual, $value);
    return 8;
} catch (ComparisonFailure $failure) {
    return false;
}
```

Instead of :
```php
try {
    $comparator->assertEquals($actual, $this->value);
    return 8;
 } catch (ComparisonFailure $failure) {
    return false;
}
```

Happy to submit PR if this is the case.

**Comment by jon-acker** [comment]
On second thoughts might be cleaner to have a new ProphecyComparator extending ObjectComparator which accepts ProphecyInterface and whose assertEquals method calls reveal() on the $actual

**Comment by everzet** [comment]
Should be closed with #234
